presto-spark-base/src/main/java/com/facebook/presto/spark/planner/PrestoSparkStatsCalculator.java

```diff
@@ -0,0 +1,92 @@
+import com.facebook.presto.spi.plan.PlanNode;
+import com.facebook.presto.spi.statistics.HistoryBasedSourceInfo;
+import com.facebook.presto.sql.planner.TypeProvider;
+import com.facebook.presto.sql.planner.iterative.Lookup;
+import com.facebook.presto.sql.planner.optimizations.PlanNodeSearcher;
+import com.facebook.presto.sql.planner.plan.RemoteSourceNode;
+
+import java.util.List;
+
+import static com.facebook.presto.cost.HistoricalPlanStatisticsUtil.similarStats;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * This StatsCalculator incorporates runtime statistics when available
+ * and decides whether to use historical or runtime-based statistics.
+ */
+public class PrestoSparkStatsCalculator
+        implements StatsCalculator
+
+{
+    private final HistoryBasedPlanStatisticsCalculator historyBasedPlanStatisticsCalculator;
+    private final StatsCalculator delegate;
+    private final HistoryBasedOptimizationConfig historyBasedOptimizationConfig;
+
+    public PrestoSparkStatsCalculator(HistoryBasedPlanStatisticsCalculator historyBasedPlanStatisticsCalculator, StatsCalculator delegate, HistoryBasedOptimizationConfig historyBasedOptimizationConfig)
+    {
+        this.historyBasedPlanStatisticsCalculator = requireNonNull(historyBasedPlanStatisticsCalculator, "historyBasedPlanStatisticsCalculator is null");
+        this.delegate = requireNonNull(delegate, "delegate is null");
+        this.historyBasedOptimizationConfig = requireNonNull(historyBasedOptimizationConfig, "historyBasedOptimizationConfig");
+    }
+
+    @Override
+    public PlanNodeStatsEstimate calculateStats(PlanNode node, StatsProvider sourceStats, Lookup lookup, Session session, TypeProvider types)
+    {
+        boolean shouldUseHistoricalStats = shouldUseHistoricalStats(node, sourceStats, lookup, session, types);
+        if (shouldUseHistoricalStats) {
+            return historyBasedPlanStatisticsCalculator.calculateStats(node, sourceStats, lookup, session, types);
+        }
+
+        return delegate.calculateStats(node, sourceStats, lookup, session, types);
+    }
+
+    private boolean shouldUseHistoricalStats(PlanNode node, StatsProvider sourceStats, Lookup lookup, Session session, TypeProvider types)
+    {
+        // RemoteSourceNode stats are computed at runtime.  If there are any
+        // RemoteSourceNodes, check whether we should use historical or runtime stats
+        // by comparing whether the runtime stats are similar to the historical stats
+        // if they are similar, use historical stats. if they differ, use runtime stats.
+        List<RemoteSourceNode> remoteSourceNodes = PlanNodeSearcher.searchFrom(node, lookup)
+                .where(RemoteSourceNode.class::isInstance)
+                .findAll();
+        for (RemoteSourceNode remoteSourceNode : remoteSourceNodes) {
+            PlanNodeStatsEstimate historicalStats = historyBasedPlanStatisticsCalculator.calculateStats(remoteSourceNode, sourceStats, lookup, session, types);
```

Comment:
Have you checked whether hbo stats come up correctly for a remotesourcenode? It may not have a statsequivalentplannode attached

Reply:
great catch. it wasn't getting populated correctly.  Didn't notice since i added it explicitly in my test.

Reply:
I put the fix for this earlier in the stack, so commit 6 is now commit 7